docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,454 @@
+and compares them with a verified header from the primary. If there is a
+conflict, it attempts to verify the conflicting header via the verifier. If it
+can be verified, it indicates an attack on the lite clients that should be
+punishable. The relevant information (ie. the two conflicting commits) are
+passed to the publisher.
+
+### Publisher
+
+For now, the publisher just logs an error, write the conflicting commits to a
+file, and exits. We leave it to a future document to describe how this
+information can actually be published to the blockchain so the validators can be
+punished. Tendermint may need to expose a new RPC endpoint to facilitate this.
+
+### Address Book
+
+For now this is a simple list of HTTPS addresses corresponding to full nodes
+that the node connects to. One is randomly selected to be the primary, while
+others serve as backups. It's essential that the lite node connect to at least
+one correct full node in order to detect conflicts in a timely fashion. We keep
+this mechanism simple for now, but in the future a more advanced peer discovery
+mechanism may be utilized.
+
+### Verifier
+
+Most of the lite client logic resides in the verifier, where commits for a
+header are actually verified. In order to abstract over all the data in the
+Tendermint data structures, we use a set of traits that include only the
+information absolutely necessary for the lite client. From this perspective, we
+have the following traits.
+
+#### Header
+
+A Header must contain a height, time, and validator sets. It can be uniquely
+identified by its hash:
+
+```rust
+pub trait Header {
+    fn height(&self) -> Height;
+    fn bft_time(&self) -> Time;
+    fn validators_hash(&self) -> Hash;
+    fn next_validators_hash(&self) -> Hash;
+
+    fn hash(&self) -> Hash;
+}
+```
+
+#### Validator
+
+A validator contributes a positive voting power if a message was correctly signed by it,
+otherwise it contributes 0. We could represent this with a single method that
+returns either 0 or the voting power, but it's probably clearer with two methods:
+
+```rust
+pub trait Validator {
+    fn power(&self) -> u64;
+    fn verify_signature(&self, sign_bytes: &[u8], signature: &[u8]) -> bool;
+}
+```
+
+#### Vote
+
+A vote corresponds to the message signed by a particular validator for a particular commit.
+Since it is expected to be for a particular commit, it does not need to include
+that information explicitly (eg. the height, the round, the block ID, etc.).
+Instead, from the verifier's perspective, all that information can be contained implicitly within the message
+bytes that are being signed. When those message bytes and the signature are passed to 
+`validator.verify_signature` for the corresponding validator, it should return true:
+
+```rust
+pub trait Vote {
+    fn validator_id(&self) -> ValID;
+    fn sign_bytes(&self) -> &[u8];
+    fn signature(&self) -> &[u8];
+}
+```
```

Comment:
Do we need to add a comment here how sign_bytes can work without having a chainId parameter?